lib/charms/operator_libs_linux/v1/systemd.py

```diff
@@ -66,127 +70,65 @@
+        [*cmd],
+        stdout=subprocess.PIPE,
+        stderr=subprocess.STDOUT,
+        text=True,
+        bufsize=1,
+        encoding="utf-8"
+    )
+    logger.debug(proc.stdout)
+    if proc.returncode == 1 and "is-failed" in cmd:
         return False
-
-    if sub_cmd == "is-failed":
+    elif proc.returncode == 3 and "is-active" in cmd:
         return False
-
-    raise SystemdError(
-        "Could not {}{}: systemd output: {}".format(
-            sub_cmd, " {}".format(service_name) if service_name else "", last_line
-        )
-    )
-
-
-def service_running(service_name: str) -> bool:
-    """Determine whether a system service is running.
-
-    Args:
-        service_name: the name of the service to check
-    """
-    return _systemctl("is-active", service_name, quiet=True)
-
-
-def service_failed(service_name: str) -> bool:
-    """Determine whether a system service has failed.
-
-    Args:
-        service_name: the name of the service to check
-    """
-    return _systemctl("is-failed", service_name, quiet=True)
-
-
-def service_start(service_name: str) -> bool:
-    """Start a system service.
-
-    Args:
-        service_name: the name of the service to start
-    """
-    return _systemctl("start", service_name)
-
-
-def service_stop(service_name: str) -> bool:
-    """Stop a system service.
-
-    Args:
-        service_name: the name of the service to stop
-    """
-    return _systemctl("stop", service_name)
-
-
-def service_restart(service_name: str) -> bool:
-    """Restart a system service.
-
-    Args:
-        service_name: the name of the service to restart
-    """
-    return _systemctl("restart", service_name)
+    elif proc.returncode == 0:
+        return True
+    else:
+        raise SystemdError(f"Command {cmd} failed. systemctl output:\n{proc.stdout}")
+
+
+service_running = functools.partial(_systemctl, "--quiet", "is-active")
+service_running.__doc__ = "Determine whether a system service is running."
+service_failed = functools.partial(_systemctl, "--quiet", "is-failed")
+service_failed.__doc__ = "Determine whether a system service has failed."
+service_start = functools.partial(_systemctl, "start")
+service_start.__doc__ = "Start a system service."
+service_stop = functools.partial(_systemctl, "stop")
+service_stop.__doc__ = "Stop a system service."
+service_restart = functools.partial(_systemctl, "restart")
+service_restart.__doc__ = "Restart a system service."
+service_enable = functools.partial(_systemctl, "enable")
+service_enable.__doc__ = "Enable a system service."
+service_disable = functools.partial(_systemctl, "disable")
+service_disable.__doc__ = "Disable a system service."
+daemon_reload = functools.partial(_systemctl, "daemon-reload")
+daemon_reload.__doc__ = "Reload systemd manager configuration."
 
 
 def service_reload(service_name: str, restart_on_failure: bool = False) -> bool:
     """Reload a system service, optionally falling back to restart if reload fails.
```

Comment:
Related to the above, what does the return value of this function mean? Or will it always return True or raise an error?

Reply:
Which function are we talking about? My view on GitHub shows multiple functions unfortunately... If we are discussing `service_reload(...)` the return value is True and an error will be raised if `systemctl reload ...` and/or `systemctl restart ...` fail (returncode != 0). True just means that the operation to reload and/or restart a service using systemctl was successful. 

This bool value returned by the `service_reload(...)` function is a holdover from the original implementation of `_systemctl(...)` where False was returned if the "is-active" or "is-failed" did not return the expected exit code to indicate that the targeted service is not running or has failed.

Reply:
Ripped off the band-aid :adhesive_bandage: and changed annotations for methods that should not return anything to `None` in https://github.com/canonical/operator-libs-linux/pull/106/commits/af8c1fb3b34fbfc19b74b20d4210dd45c7eb07db.

Now, if an error is encountered, a `SystemdError` will be raised. The returned exit code of `0` by _systemctl is ignored in the public methods, but _systemctl itself will log the exit code, command, and stdout when in debug mode.